posts/2021/08/better-interractive-jupyter-session-with-pyflyby.md

```diff
@@ -0,0 +1,202 @@
++++ example.py	2021-03-08 10:33:18.000000000 -0800
+@@ -1,2 +1,7 @@
++from   matplotlib               import pyplot
++import pandas as pd
++
+ data = pd.read_csv("./data/base-pop-2015.csv")
+ pyplot.plot(data.population)
+
+Replace example.py? [y/N]
+```
+
+## Other utilities
+
+Pyflyby contains a number of other utilities to improve the efficiency of
+manipulating and executing Python code. (Please refer to pyflyby's
+[README](https://github.com/deshaw/pyflyby) for additional information.)
+
+`py` is one such example. It is a flexible tool that can be used either to start
+IPython or to execute commands quickly from the developer's shell without the
+need for imports. It supports a range of syntax options, allowing for quick
+calculation and graph plotting.
+
+ - Without any parameters, `py` will start IPython with the pyflyby extension
+   activated.
+ - With space-separated arguments, `py` will attempt to interpret each argument
+   as a Python function call with the right imports:
+
+
+```bash
+$ py np.random.normal 0 1
+[PYFLYBY] import numpy as np
+[PYFLYBY] np.random.normal(0, 1)
+-0.027577422117386
+```
+
+- When more control over values is necessary, `py` will run a Python expression:
+
+```bash
+$ py 'plot(scipy.stats.norm.pdf(linspace(-5, 5), 0, 1))'
+[PYFLYBY] from numpy import linspace
+[PYFLYBY] from matplotlib.pyplot import plot
+[PYFLYBY] import scipy.stats
+[PYFLYBY] plot(scipy.stats.norm.pdf(linspace(-5, 5), 0, 1))
+[<matplotlib.lines.Line2D object at 0x132981940>]
+```
+
+![using pyflyby from bash to plot with matplotlib with above snippet, at matplotlib widow is open and show the plot.](/images/2021/08/py-exec-matplotlib.png)
+
+`find-import` , another utility available in pyflyby, can be deployed to find a
+particular function across many libraries by returning the relevant import.  For
+example:
+
+```bash
+$ find-import norm
+from scipy.stats.distributions import norm
+```
+
+# Notes on pyflyby codebase
+
+The [Pyflyby codebase](https://github.com/deshaw/pyflyby) provides a window into
+advanced data structures and programming paradigms. Its use of modules and
+programming concepts are unusual relative to those found in more classical data
+science-focused libraries. For example:
+
+ - Pyflyby will conduct non-trivial manipulation of the Python Abstract Syntax
+   Tree (AST), which represents code written in tree form. Pyflyby uses AST to
+   find and insert missing imports, and it does so even as AST's exact
+   representation changes with almost every minor release.
+
+ - Pyflyby demonstrates the use of  [Aspect-Oriented
+   programming](https://en.wikipedia.org/wiki/Aspect-oriented_programming),
+   which highlights the flexibility of the Python programming model.
+
+
+# Conclusion
+
+Pyflyby's utilities are designed to improve developer efficiency and materially
+reduce the impact of interruption on productivity. Its value, however, is
+broader than that. Pyflyby expands what is capable within the Python ecosystem
```

Comment:
```suggestion
broader than that. Pyflyby expands what one is capable of doing within the Python ecosystem
```
"what is capable" sounds a bit weird specially because it is missing a subject so I suggest "what one is capable of doing" instead but maybe something like "what we can do within the Python ecosystem...." could work too.

Reply:
That also looked weird to me, but as this was in a change D.E. Shaw suggested so I left it originally.

Will commit all your suggestions.